### DISCUSSION TITLE
Related Discord discussion -- ReArch/React tree fine integration

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#157] [General]
https://discord.com/channels/102860784329052160/103696749012467712/1232020137591308379

![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/242ff7f7-16cb-4624-83a8-027e07a89e9b)

![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/ab92afe0-ebd2-4892-87ec-d6e77a3be95e)

![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/bd7a1156-b289-4421-88ff-86edce0ff4a5)

![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/2e95fb9a-7ec2-472d-abdf-aa77f7fd6916)

![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/1906d59d-8b19-4a14-a875-0e16d72dd25d)

![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/e9f46e59-8937-4458-97b3-ea5365d7867d)

![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/09dd2dc2-5433-49c8-9ee7-dc242297fc37)

![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/e216079c-b8f7-4da9-87a9-7b0745b82c7e)

![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/49e6b210-0df6-481c-9995-af89dd3f3b82)

![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/0633d821-57b5-4841-ba8f-acd27f84bb85)

**Comment by GregoryConrad** [comment]
W.r.t. the comments on that discord, ReArch is a lot more powerful than zustand. You can use composition across states/app-level components and features/side effects. No other library I’m aware of lets you use composition like that to solve virtually every problem out there.

Actually just wrote a bit on that here: https://old.reddit.com/r/FlutterDev/comments/1c801tv/flutter_design_patterns_research_for_flutterdev/l0bkvi2/

feel free to link it in the discord 😝

**Reply by GregoryConrad** [reply]
The same is true for signals too—ReArch is similar to signals but with the ability to use a lot more powerful composition.

**Comment by busslina** [comment]
> feel free to link it in the discord 😝

Done

**Comment by busslina** [comment]
https://github.com/Workiva/react-dart/issues/394

I don't really know if `useSyncExternalStore` is the solution I need to accomplish the ReArch/React tree fine integration.

Also, it seems only available as hook (functional Component) and not for class Component.

**Reply by GregoryConrad** [reply]
I would help ya out more, but unfortunately I’m not too versed in the React space. I only touched it once for a total of 10 minutes…

**Reply by busslina** [reply]
No problem! Thanks for all

**Comment by busslina** [comment]
I'm just about to giving up with this React port... and maybe creating my HTML/JS framework (on top of [web](https://pub.dev/packages/web) instead of [html](https://pub.dev/packages/html))

**Reply by busslina** [reply]
Do you think that if I do my own framework I could store sub/tree info as ReArch state on every component?

**Reply by busslina** [reply]
In ReArch, you can only have a capsule instance per container. But, when applied to a GUI tree model, you can have multiple instance of the same component (in multiple tree locations). Does ReArch supports it well? Does it stores a different state copy for each component instance? (I suppose yes, just asking for make it fully clear)

**Reply by GregoryConrad** [reply]
You mean like a 1-1 component/widget to capsule mapping? If so, yes, because ReArch is an incremental computation engine. I might need to make a couple small API additions (to prevent leaks in select certain scenarios when a stateful widget is removed from the tree), but that should be completely doable.

You should be able to create the backing store for the view tree entirely using capsules. You’ll need “dynamic capsules,” which there is a test or two for, but not documented (at least on the Dart side, since so few folks using Dart need an IC-specific engine).

the idea is that a widget “A” that has a widget child “B” would have a backing capsule “A” and capsule “B”, where the B capsule use()s A. Thus, whenever A is rebuilt, B is also automatically rebuilt, and in the proper order. The way this works is that the underlying A/B capsule provides its CapsuleHandle to the widget itself in order to build the widget. This in turn takes care of all widget state and capsule dependencies for you for free. This acts similar to the Element tree in Flutter, but is a bit more elegant since it is all built via ReArch.

**Reply by busslina** [reply]
Cool!

**Reply by busslina** [reply]
I'm starting it as a private framework but if I achieved it I will post a PR .

**Reply by busslina** [reply]
> You’ll need “dynamic capsules,” which there is a test or two for

I don't find it. Can you link it?

Maybe they are created as inner-functions (AKA closures)...

**Reply by GregoryConrad** [reply]
On my phone, but search the ReArch package tests for “fibonacci”

**Reply by busslina** [reply]
Wow, Fibonacci test is wonderful!

**Reply by busslina** [reply]
So maybe no Component handle implementation is needed. Just using CapsuleHandle.

A root capsule + dynamic capsules, all in one.

A capsule renders his node on the first build. It replaces it on the next rebuilds and it removes it on dispose. Using use effect.

Does it sound good to you?

**Reply by busslina** [reply]
Maybe not using use effect, but similar

**Reply by GregoryConrad** [reply]
Well I was thinking you’d have one capsule to manage all the dynamic capsules here, but then also a dynamic capsule for each view. The key of the dynamic capsule would be the view/widget key (like a Flutter GlobalKey, or if none is specified, the type & path of the view/widget).

you’d likely still want a ComponentHandle that abstracts over the CapsuleHandle; this would enable you to add things like Flutter’s context

**Reply by busslina** [reply]
I don't really get it.

Up to now, I made it work, but whenever there is a rebuild (even on a leaf) the whole tree is rebuilding. I don't know why, maybe it's because I'm using the same `ComponentHandle` (the root component one) for all the tree.

Can you take a look on it (see example/web/main.dart)?

Can advice me based on this code?

https://github.com/busslina/busslina-dart-web-framework
[Video_2024_04_23-1.webm](https://github.com/GregoryConrad/rearch-dart/assets/73592852/45481059-55de-44ba-8048-cfd2a3610670)

UPDATE: now it's not rebuilding cause I change usage from CapsuleHandle to ComponentHandle (only using the root one), but I feel that I'm closing the gap...

**Reply by GregoryConrad** [reply]
The idea is:

```dart
Component aComponent(ComponentHandle use) {
  return Component()
    .key(globalKey(123))
    .child(padding(8))
    .child(bComponent); // will add the use(bCapsule) call later on
}

Component bComponent(ComponentHandle use) {
  return Component()
    .child(container(...));
}
```

Would then be evaluated into something like:

```dart
BuiltComponent aCapsule(CapsuleHandle use) {
  return aComponent(ComponentHandleImpl(use)).render();
}

// executing aCapsule would then create this dynamic capsule:
BuiltComponent bCapsule(CapsuleHandle use) {
  use(aCapsule); // rebuild when parent rebuilds
  return bComponent(ComponentHandleImpl(use)).render();
}
```

Both aCapsule and bCapsule would be closures stored under a `componentManager` capsule, much like the root fibonacci capsule that orchestrates the other fibonacci capsules. They'd be stored as a combination of their type (perhaps the function itself in this case, if a functional approach like the above example is chosen), in addition to the node path to the component in question. There are also different key types, like a GlobalKey, which would have to be accounted for.

The only issue here is that this will currently cause a leak for stateful components if they are removed from the tree, but that is an easy enough API fix that I can work on eventually here.

**Reply by GregoryConrad** [reply]
You would also need to integrate some sort of context into all of this to handle scoped values (like InheritedWidget). That could probably be done with a HashMap from fast_immutable_collections (if there is a hash map in that package).

**Reply by GregoryConrad** [reply]
If I had a good chunk of spare time, I could probably just make this framework--probably wouldn't be terrible. Bit tight on time right now though.

**Reply by busslina** [reply]
Oh, I think I got it now. Anyway, I will finish my not full FP/declarative approach and once it works fine I will try to adapt to this pattern that is just much more elegant.
